Kudu.Agent/Controllers/KuduProcessController.cs

```diff
@@ -0,0 +1,324 @@
+            ProcessThread thread = threads.FirstOrDefault(t => t.Id == threadId);
+            if (thread == null)
+            {
+                return NotFound();
+            }
+
+            ProcessThreadInfo threadInfo = GetProcessThreadInfo(thread, details: true);
+
+            return Ok(threadInfo);
+        }
+
+        [HttpGet("{id}/modules")]
+        public IActionResult GetAllModules(int id)
+        {
+            Process process = GetProcessById(id);
+            IEnumerable<ProcessModuleInfo> modulesInfo = GetModules(process);
+
+            return Ok(modulesInfo);
+        }
+
+        [HttpGet("{id}/modules/{baseAddress}")]
+        public IActionResult GetModule(int id, string baseAddress)
+        {
+            Process p = GetProcessById(id);
+            IEnumerable<ProcessModule> modules = p.Modules.Cast<ProcessModule>();
+            ProcessModule module = modules.FirstOrDefault(t => t.BaseAddress.ToInt64() == Int64.Parse(baseAddress, NumberStyles.HexNumber));
+            if (module == null)
+            {
+                return NotFound();
+            }
+
+            ProcessModuleInfo moduleInfo = GetProcessModuleInfo(module, details: true);
+
+            return Ok(moduleInfo);
+        }
+
+        [HttpGet("{id}/environments/{filter}")]
+        public IActionResult GetEnvironments(int id, string filter)
+        {
+            throw new NotImplementedException();
+        }
+
+        [HttpGet("{id}/dump")]
+        public IActionResult MiniDump(int id, int dumpType = 0, string format = null)
+        {
+            throw new NotImplementedException();
+        }
+
+        [HttpPost("{id}/profile/start")]
+        public IActionResult StartProfileAsync(int id, bool iisProfiling = false)
+        {
+            throw new NotImplementedException();
+        }
+
+        [HttpGet("{id}/profile/stop")]
+        public IActionResult StopProfileAsync(int id)
+        {
+            throw new NotImplementedException();
+        }
+
+
+
+        private static Process GetProcessById(int processId)
+        {
+            if (processId == 0)
+            {
+                return Process.GetCurrentProcess();
+            }
+            else if (processId == -1)
+            {
+                // Order is important
+                string[] processesOfInterest = new string[]
+                {
+                    "w3wp",
+                    "dotnet",
+                    "java",
+                    "node"
+                };
+
+                foreach (string processOfInterest in processesOfInterest)
+                {
+                    Process p = Process.GetProcessesByName(processOfInterest).FirstOrDefault();
+                    if (p != null && !AppServicePlatformProcess(p))
+                    {
+                        return p;
+                    }
+                }
+
+                return null;
+            }
+
+            return Process.GetProcessById(processId);
```

Comment:
It would add more sense to return NotFound when a process by a given Id is not found in that case this Process.GetProcessById(processId) would throw an exception.

Reply:
I think I will leave it as returning null and allow its callers to handle the null checking. Returning NotFound would mean I'd have to change the function to return an object that can encapsulate either a Process or an IActionResult. I think that might introduce unnecessary complexity.